bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -2338,33 +2338,37 @@ private static boolean checkIsLikeArrayType(Object sourceValue, BArrayType targe
         if (source.getType().getTag() == TypeTags.ARRAY_TAG) {
             Type sourceElementType = ((BArrayType) source.getType()).getElementType();
             if (isValueType(sourceElementType)) {
-                boolean isType = checkIsType(sourceElementType, targetTypeElementType, new ArrayList<>());
 
-                if (isType || !allowNumericConversion || !isNumericType(sourceElementType)) {
-                    return isType;
-                }
-
-                if (isNumericType(targetTypeElementType)) {
+                if (checkIsType(sourceElementType, targetTypeElementType, new ArrayList<>())) {
                     return true;
                 }
 
-                if (targetTypeElementType.getTag() != TypeTags.UNION_TAG) {
-                    return false;
-                }
+                if (allowNumericConversion && isNumericType(sourceElementType)) {
+                    if (isNumericType(targetTypeElementType)) {
+                        return true;
+                    }
 
-                List<Type> targetNumericTypes = new ArrayList<>();
-                for (Type memType : ((BUnionType) targetTypeElementType).getMemberTypes()) {
-                    if (isNumericType(memType) && !targetNumericTypes.contains(memType)) {
-                        targetNumericTypes.add(memType);
+                    if (targetTypeElementType.getTag() != TypeTags.UNION_TAG) {
+                        return false;
                     }
+
+                    List<Type> targetNumericTypes = new ArrayList<>();
+                    for (Type memType : ((BUnionType) targetTypeElementType).getMemberTypes()) {
+                        if (isNumericType(memType) && !targetNumericTypes.contains(memType)) {
+                            targetNumericTypes.add(memType);
+                        }
+                    }
+                    return targetNumericTypes.size() == 1;
+                }
+
+                if (isNumericType(targetTypeElementType) && targetTypeElementType.getTag() != TypeTags.BYTE_TAG) {
```

Comment:
Why are special casing byte here?

Reply:
For early return of numeric target type arrays except byte arrays, where the `allowNumericConversion` flag is false, where we do not need to check element wise. For example casting of (readonly) `float` array to `int` array where we do not need to check element by element.

Reply:
would this case fail for `int:Unsigned16` ?

Reply:
Since the check is considering the types which have the type tag values less than 5 (which is the value of `STRING_TAG`) as `numeric types`, that condition will not stop sub types which have type tag values >= 40. It will work for all subtypes as expected. 
e.g:
for `Unsigned16`
```
public function main() {
    (readonly & int[]) a = [1, 2, 6553];
    any b =  a;
    int:Unsigned16[] c = <int:Unsigned16[]> b; 
}
```
will be valid casting and 
```
public function main() {
    (readonly & int[]) a = [1, 2, 65536];
    any b =  a;
    int:Unsigned16[] c = <int:Unsigned16[]> b; 
}
``` 
will result in panic. 

This case will omit the unnecessary checks for `float` and `decimal` types.